include/aws/crt/cbor/Cbor.h

```diff
@@ -0,0 +1,362 @@
+
+                /**
+                 * Encode a simple value Break
+                 * Notes: no error checking, it's user's responsibility to track the break to close the corresponding
+                 * indef_start
+                 */
+                void WriteBreak() noexcept;
+
+                /**
+                 * Encode an IndefBytesStart
+                 * Notes: no error checking, it's user's responsibility to add corresponding data and the break to close
+                 * the indef_start
+                 */
+                void WriteIndefBytesStart() noexcept;
+
+                /**
+                 * Encode an IndefTextStart
+                 * Notes: no error checking, it's user's responsibility to add corresponding data and the break to close
+                 * the indef_start
+                 */
+                void WriteIndefTextStart() noexcept;
+
+                /**
+                 * Encode an IndefArrayStart
+                 * Notes: no error checking, it's user's responsibility to add corresponding data and the break to close
+                 * the indef_start
+                 */
+                void WriteIndefArrayStart() noexcept;
+
+                /**
+                 * Encode an IndefMapStart
+                 * Notes: no error checking, it's user's responsibility to add corresponding data and the break to close
+                 * the indef_start
+                 */
+                void WriteIndefMapStart() noexcept;
+
+              private:
+                struct aws_cbor_encoder *m_encoder;
+            };
+
+            class AWS_CRT_CPP_API CborDecoder final
+            {
+
+              public:
+                CborDecoder(const CborDecoder &) = delete;
+                CborDecoder(CborDecoder &&) = delete;
+                CborDecoder &operator=(const CborDecoder &) = delete;
+                CborDecoder &operator=(CborDecoder &&) = delete;
+
+                /**
+                 * Construct a new Cbor Decoder object
+                 *
+                 * @param allocator
+                 * @param src The src data to decode from.
+                 */
+                CborDecoder(Allocator *allocator, ByteCursor src) noexcept;
+                ~CborDecoder() noexcept;
+
+                /**
+                 * Get the length of the remaining bytes of the source. Once the source was decoded, it will be
+                 * consumed, and result in decrease of the remaining length of bytes.
+                 *
+                 * @return The length of bytes remaining of the decoder source.
+                 */
+                size_t GetRemainingLength() noexcept;
+
+                /**
+                 * Decode the next element and store it in the decoder cache if there was no element cached.
+                 * If there was an element cached, just return the type of the cached element.
+                 *
+                 * @param out_type
+                 * @return success/failure
```

Comment:
document what the user should do in the case of failure. We never really settled on a standardized way to do last-error in aws-crt-cpp, so you need to tell people how THIS class handles it